File: packages/vscode-extension/src/extension/session/providers.ts
import { type Uri, workspace } from "#vscode-host";
import {
  dependencyPropertyKeys,
  filePatternKeys,
  prereleaseTagKeys,
  providerCacheKeys,
  providerStrictSslKeys,
  registryUrlKeys,
} from "../config/keys.ts";
import type {
  NativeDependencyPropertyConfig,
  NativeFilePatternConfig,
  NativePrereleaseTagFilter,
  NativeProviderCacheConfig,
  NativeProviderHttpConfig,
  NativeRegistryUrl,
} from "../native.ts";
import { configuredValue } from "./configured.ts";

export function dependencyProperties(
  resource?: Uri,
): NativeDependencyPropertyConfig[] {
  const config = workspace.getConfiguration("versionlens", resource);

  return dependencyPropertyKeys.flatMap(
    ([ecosystem, key]): Array<{
      ecosystem: string;
      properties: string[];
    }> => {
      const properties = configuredValue<string[] | undefined>(key, config);
      if (properties === undefined) {
        return [];
      }
      return [{ ecosystem, properties }];
    },
  );
}

export function filePatterns(resource?: Uri): NativeFilePatternConfig[] {
  const config = workspace.getConfiguration("versionlens", resource);

  return filePatternKeys.flatMap(
    ([ecosystem, key]): Array<{ ecosystem: string; pattern: string }> => {
      const pattern = configuredValue<string | undefined>(key, config);
      if (pattern === undefined) {
        return [];
      }
      return [{ ecosystem, pattern }];
    },
  );
}

export function registryUrls(resource?: Uri): NativeRegistryUrl[] {
  const config = workspace.getConfiguration("versionlens", resource);
  const configuredRegistryUrls = registryUrlKeys.flatMap(
    ([ecosystem, key]): Array<{ ecosystem: string; url: string }> => {
      const url = configuredValue<string | undefined>(key, config);
      if (url === undefined) {
        return [];
      }
      return [{ ecosystem, url }];
    },
  );

  for (const url of configuredValue<(string | undefined)[] | undefined>(
    "dotnet.nuget.sources",
    config,
  ) ?? []) {
    if (url !== undefined) {
      configuredRegistryUrls.push({ ecosystem: "dotnet", url });
    }
  }

  return configuredRegistryUrls;
}

export function prereleaseTagFilters(
  resource?: Uri,
): NativePrereleaseTagFilter[] {
  const config = workspace.getConfiguration("versionlens", resource);

  return prereleaseTagKeys.flatMap(
    ([ecosystem, key]): Array<{ ecosystem: string; tags: string[] }> => {
      const tags = configuredValue<string[] | undefined>(key, config);
      if (tags === undefined) {
        return [];
      }
      return [{ ecosystem, tags }];
    },
  );
}

export function providerHttpConfigs(
  resource?: Uri,
): NativeProviderHttpConfig[] {
  const config = workspace.getConfiguration("versionlens", resource);

  return providerStrictSslKeys.flatMap(
    ([ecosystem, key]): Array<{ ecosystem: string; strictSsl: boolean }> => {
      const strictSsl = configuredValue<boolean | null | undefined>(
        key,
        config,
      );
      if (strictSsl === undefined || strictSsl === null) {
        return [];
      }
      return [{ ecosystem, strictSsl }];
    },
  );
}

export function providerCacheConfigs(
  resource?: Uri,
): NativeProviderCacheConfig[] {
  const config = workspace.getConfiguration("versionlens", resource);

  return providerCacheKeys.flatMap(
    ([ecosystem, key]): Array<{
      cacheDurationMinutes: number;
      ecosystem: string;
    }> => {
      const cacheDurationMinutes = configuredValue<number | null | undefined>(
        key,
        config,
      );
      if (cacheDurationMinutes === undefined || cacheDurationMinutes === null) {
        return [];
      }
      return [{ cacheDurationMinutes, ecosystem }];
    },
  );
}

File: crates/versionlens-napi/src/model/config/session/tests/providers.rs
use super::blank_session_config;
use crate::model::config::{
    NativeDependencyPropertyConfig, NativeFilePatternConfig, NativeProviderCacheConfig,
    NativeProviderHttpConfig, NativeProviderSettings, NativeRegistryUrl, NativeSessionConfig,
};
use versionlens_parsers::Ecosystem::{Dotnet, Npm};
use versionlens_parsers::ManifestKind::{ComposerJson, PnpmYaml};

#[test]
fn pnpm_config_namespace_maps_to_npm_provider() {
    let config = NativeSessionConfig {
        enabled_providers: Some(vec!["pnpm".to_owned()]),
        providers: Some(NativeProviderSettings {
            registry_urls: None,
            prerelease_tag_filters: None,
            provider_cache: Some(vec![NativeProviderCacheConfig {
                ecosystem: "pnpm".to_owned(),
                cache_duration_minutes: Some(2.0),
            }]),
            provider_http: Some(vec![NativeProviderHttpConfig {
                ecosystem: "pnpm".to_owned(),
                strict_ssl: Some(false),
            }]),
            dependency_properties: Some(vec![NativeDependencyPropertyConfig {
                provider: None,
                ecosystem: "pnpm".to_owned(),
                properties: vec!["catalog".to_owned()],
            }]),
            file_patterns: None,
        }),
        ..blank_session_config()
    }
    .into_core();

    assert_eq!(config.enabled_providers[0].ecosystem, Npm);
    assert_eq!(config.enabled_providers[0].manifest_kind, Some(PnpmYaml));
    assert_eq!(config.providers.provider_cache[0].ecosystem, Npm);
    assert_eq!(
        config.providers.provider_cache[0].manifest_kind,
        Some(PnpmYaml)
    );
    assert_eq!(config.providers.provider_cache[0].cache_ttl_ms, 120_000);
    assert_eq!(config.providers.provider_http[0].ecosystem, Npm);
    assert_eq!(
        config.providers.provider_http[0].manifest_kind,
        Some(PnpmYaml)
    );
    assert_eq!(config.providers.provider_http[0].strict_ssl, Some(false));
    assert_eq!(config.providers.dependency_properties[0].ecosystem, Npm);
    assert_eq!(
        config.providers.dependency_properties[0].manifest_kind,
        Some(PnpmYaml)
    );
}

#[test]
fn provider_fractional_cache_minutes_are_converted_to_milliseconds() {
    let config = NativeSessionConfig {
        providers: Some(NativeProviderSettings {
            registry_urls: None,
            prerelease_tag_filters: None,
            provider_cache: Some(vec![NativeProviderCacheConfig {
                ecosystem: "npm".to_owned(),
                cache_duration_minutes: Some(0.25),
            }]),
            provider_http: None,
            dependency_properties: None,
            file_patterns: None,
        }),
        ..blank_session_config()
    }
    .into_core();

    assert_eq!(config.providers.provider_cache[0].cache_ttl_ms, 15_000);
}

#[test]
fn registry_urls_are_trimmed_and_blank_urls_are_ignored_in_rust() {
    let config = NativeSessionConfig {
        providers: Some(NativeProviderSettings {
            registry_urls: Some(vec![
                NativeRegistryUrl {
                    ecosystem: "cargo".to_owned(),
                    url: "   ".to_owned(),
                },
                NativeRegistryUrl {
                    ecosystem: "cargo".to_owned(),
                    url: " https://mirror.test/crates ".to_owned(),
                },
                NativeRegistryUrl {
                    ecosystem: "dotnet".to_owned(),
                    url: " https://configured.nuget/v3/index.json ".to_owned(),
                },
            ]),
            prerelease_tag_filters: None,
            provider_cache: None,
            provider_http: None,
            dependency_properties: None,
            file_patterns: None,
        }),
        ..blank_session_config()
    }
    .into_core();

    assert_eq!(config.providers.registry_urls.len(), 2);
    assert_eq!(
        config.providers.registry_urls[0].url,
        "https://mirror.test/crates"
    );
    assert_eq!(
        config
            .providers
            .registry_urls
            .iter()
            .filter(|url| url.ecosystem == Dotnet)
            .map(|url| url.url.as_str())
            .collect::<Vec<_>>(),
        ["https://configured.nuget/v3/index.json"]
    );
}

#[test]
fn file_patterns_are_trimmed_and_mapped_to_manifest_kinds() {
    let config = NativeSessionConfig {
        providers: Some(NativeProviderSettings {
            registry_urls: None,
            prerelease_tag_filters: None,
            provider_cache: None,
            provider_http: None,
            dependency_properties: None,
            file_patterns: Some(vec![
                NativeFilePatternConfig {
                    ecosystem: "composer".to_owned(),
                    pattern: " **/acme.composer.json ".to_owned(),
                },
                NativeFilePatternConfig {
                    ecosystem: "composer".to_owned(),
                    pattern: " ".to_owned(),
                },
            ]),
        }),
        ..blank_session_config()
    }
    .into_core();

    assert_eq!(config.providers.file_patterns.len(), 1);
    assert_eq!(
        config.providers.file_patterns[0].manifest_kind,
        ComposerJson
    );
    assert_eq!(
        config.providers.file_patterns[0].pattern,
        "**/acme.composer.json"
    );
}
